### ISSUE TITLE
Running on Specific Devices

### ISSUE THREAD

**Issue by fatahillahardhi** [prashant-ramcharan/courgette-jvm#391]
Hello Prashant, I'd like to execute a somewhat specific scenario.

I have a test on iOS with a scenario that involves capturing with a camera, and it can only be done on a Real Device. However, currently, I'm running the test on both a Real Device and a Simulator simultaneously. Now, I want to ensure that when running a scenario containing the tags `@UploadPhotos`, it should only be executed on a Real Device and not on the Simulator. Is there a solution for this? Thank you.

**Comment by prashant-ramcharan** [comment]
Hi @fatahillahardhi , 

This currently is not supported however I can update Courgette to either return a simulator device id or a real device id based on whichever property you define in your test.

I will look at implementing this.

**Comment by prashant-ramcharan** [comment]
Hi @fatahillahardhi 

As this is not currently supported, I will look into implementing it to be able to differentiate between simulators and real devices in the same run.

**Comment by fatahillahardhi** [comment]
Hi @prashant-ramcharan , hope you're doing well,

Thank you for the answer. I'm very interested to wait for the updates!

**Comment by prashant-ramcharan** [comment]
@fatahillahardhi This is now supported in Courgette version `6.11.0`

**Usage example**

_Courgette will only allocate a real device to tests tagged with `@UploadPhotos`. For all other tests, one of the simulators will be allocated._

````java
@RunWith(Courgette.class)
@CourgetteOptions(
        ...
        plugin = {CourgettePlugin.MOBILE_DEVICE_ALLOCATOR},
        mobileDeviceType = MobileDeviceType.SIMULATOR_AND_REAL_DEVICE,
        mobileDevice = {
                "simulator1",
                "simulator2"
                "realDevice:00000000-000-0000-0000-000000000001",
        },
        realMobileDeviceTag = {"@UploadPhotos"},
)
````

**Comment by fatahillahardhi** [comment]
Hi @prashant-ramcharan, it's really awesome!

However, I have encountered a problem. When I try to run it with `SIMULATOR_AND_REAL_DEVICES`, it always runs on Real Devices.

Here are my Test Runner:
```
@RunWith(Courgette.class)
@CourgetteOptions(
        runLevel = CourgetteRunLevel.SCENARIO,
        reportTargetDir = "build",
        testOutput = CourgetteTestOutput.CONSOLE,
        environmentInfo = "project=iOS Test",
        plugin = {CourgettePlugin.MOBILE_DEVICE_ALLOCATOR},
        mobileDeviceType = MobileDeviceType.SIMULATOR_AND_REAL_DEVICE,
        mobileDevice = {
                "iPhone 14",
                "iPhone Real:xxxxxxx-xxxxxxx"
        },
        realMobileDeviceTag = {"@UploadPhotos"},
        cucumberOptions = @CucumberOptions(
                features = "src/test/resources/features",
                publish = true,
                plugin = {
                        "pretty",
                        "json:build/cucumber.json"}
        ))
```

When I run the command: `./gradlew clean test -Dcourgette.threads=2 -Dcucumber.tags="@focus" -Dcourgette.rerunFailedScenarios=false`, it consistently runs on Real Devices.

Moreover, when I try to print `CourgetteMobileDeviceAllocator.DEVICE_NAME` to the terminal, it shows "iPhone 14," but in reality, the execution occurs on Real Devices, and it doesn't run in parallel.

**Comment by prashant-ramcharan** [comment]
Hi, if you remove the `realMobileDeviceTag` option from your runner, does all test run on the simulator?

Also, could you please share how your `@UploadPhotos` scenario is tagged? 

Is it tagged on scenario level, scenario outline (example table) level or at the feature level?

**Comment by fatahillahardhi** [comment]
Hi @prashant-ramcharan,

I have tried several tests with several cases:

Devices:
- 1 Simulator
- 1 Real Device

Case 1: Run with 2 scenarios, one of which contains the @UploadPhotos tag.
Results:
- Automation test does not run in parallel ❌
- Runs on both Simulator and Real Devices ✅
- Scenario with the @UploadPhotos tag does not run on the Simulator ✅

Case 2: Run with 3 scenarios, one of which contains the @UploadPhotos tag, and remove the realMobileDeviceTag.
Results:
- Automation test does not run in parallel ❌
- Only runs on the Simulator ❌
- Scenario with the @UploadPhotos tag runs on the Simulator ❌

Case 3: Run with 3 scenarios without the @UploadPhotos tag and remove the realMobileDeviceTag.
Results:
- Automation test does not run in parallel ❌
- Only runs on the Simulator ❌
- Scenario with the @UploadPhotos tag runs on the Simulator ❌


Overall I think `realMobileDeviceTag` is doing well, but the issues is the test do not run in parallel but instead run sequentially, similar to conventional testing.


Scenario:
```
@iOS @Example

  Feature: Scenario with Upload Tags

    @Test1
    Scenario: Login with already registered phone number
      Given User is on screen
      When User click button login
      And User insert phone number "0823141284"
      And User input correct OTP
      Then User is on bibit main page

    @Test2 @UploadPhotos
    Scenario: Login with unregistered phone number
      Given User is on screen
      When User click button login
      And User insert phone number unregistered "081231231231"
      Then User direct otp page without get the otp

    @Test3 @UploadPhotos
    Scenario: Login with unregistered phone number 2
      Given User is on screen
      When User click button login
      And User insert phone number unregistered "081231231231"
      Then User direct otp page without get the otp
```

Test Runner
```
@RunWith(Courgette.class)
@CourgetteOptions(
        threads = 2,
        runLevel = CourgetteRunLevel.SCENARIO,
        rerunFailedScenarios = true,
        rerunAttempts = 1,
        reportTargetDir = "build",
        testOutput = CourgetteTestOutput.CONSOLE,
        environmentInfo = "project=iOS Test",
        plugin = {CourgettePlugin.MOBILE_DEVICE_ALLOCATOR},
        mobileDeviceType = MobileDeviceType.SIMULATOR_AND_REAL_DEVICE,
        mobileDevice = {
                "iPhone:xxxxxxx-xxxxxxx",
                "iPhone 14"
        },
        realMobileDeviceTag = {"@UploadPhotos"},
        cucumberOptions = @CucumberOptions(
                features = "src/test/resources/features",
                tags = "@Example",
                publish = true,
                plugin = {
                        "pretty",
                        "json:build/cucumber.json"}
        ))
```

**Comment by prashant-ramcharan** [comment]
Hi, with regards to this point:

_Automation test does not run in parallel_

You would need to specify more than 1 simulator or device to be able to run in parallel as Courgette can only allocate 1 device per run at a time.

We wont be able to run multiple tests in parallel if we only have 1 device / simulator per run.

Example:

````java
 threads = 3,
 mobileDevice = {
                "iPhone:xxxxxxx-xxxxxxx",
                "iPhone 14",
                "iPhone 15"
                "iPhone 15 Pro"
        },
````

The above will run 3 tests in parallel on the `simulators` and only run tests on the real device if it matches the `realMobileDeviceTag`

Did you try adding more simulators / devices to see it runs in parallel?

Could you also try running the example project and let me know how you get on?

https://github.com/prashant-ramcharan/courgette-jvm-appium-ios

**Comment by fatahillahardhi** [comment]
Hi @prashant-ramcharan,

I have just tried adding more devices, but the result remains the same

_Automation test does not run in parallel_

 ```
mobileDeviceType = MobileDeviceType.SIMULATOR_AND_REAL_DEVICE,
mobileDevice = {
                "iPhone 14",
                "iPhone 1:00008020-00146850340B402E",
                "iPhone 2:00008020-001531EA2E22402E",
                "iPhone 4:00008030-001419043428C02E",
                "iPhone 6:3c240073ed0e0ecf6f786ebed31a9ee0b147e8bf",
},
realMobileDeviceTag = {"@UploadPhotos"}
```

**Comment by prashant-ramcharan** [comment]
Could you please try using just simulators to test?

Are you able to provide a demo project to reproduce this issue?

**Comment by prashant-ramcharan** [comment]
Hi @fatahillahardhi , no worries 🙂

I'm also using `MobileDeviceType.SIMULATOR_AND_REAL_DEVICE` and it runs fine in parallel. 

Please see video demo [here](https://www.dropbox.com/scl/fi/fpeo7egzk1nxorvv2o8og/issue_391_2.mov?rlkey=ymo21ovmprhvmwcmjz2quj1dd&dl=0)

As you can see in the video, the tests ran in parallel in the simulators and failed when running on the real device because the device did not exist.

If you want to share your Appium logs during the run, I can take a look. Might be worth re-building / signing your web driver agent on the real devices.

Not sure how to reproduce this, so please provide a demo project if possible.

**Comment by fatahillahardhi** [comment]
Hi @prashant-ramcharan,

I attempted to run the example project [here](https://github.com/prashant-ramcharan/courgette-jvm-appium-ios) and encountered a unique result.

I used both `MobileDeviceType.SIMULATOR` and `MobileDeviceType.SIMULATOR_AND_REAL_DEVICE`. Surprisingly, the results were the same whether I ran the tests with 3 simulators or with 3 simulators + 1 real device. Unfortunately, the automation tests did not run in parallel.

Case 1, when using either `MobileDeviceType.SIMULATOR` or `MobileDeviceType.SIMULATOR_AND_REAL_DEVICE` and running with 3 simulators or 3 simulators + 1 real device:
- The automation test was expected to run in parallel but encountered errors.
- Out of 4 scenarios, only 1 scenario passed, while the other 3 failed consistently after multiple attempts.
- You can find the logs [here](https://gist.github.com/fatahillahardhi/995cdb77a6caebf51552855a2480a186) (running with simulators only).

Case 2, when using `MobileDeviceType.SIMULATOR` and running with 1 simulator:
- All scenarios passed successfully.
- You can find the logs [here](https://gist.github.com/fatahillahardhi/055007899954b856cc78aaaa9e25f85c).

I hope this information helps in identifying and resolving the issues.


Test Runner:
```
@RunWith(Courgette.class)
@CourgetteOptions(
        threads = 3,
        runLevel = CourgetteRunLevel.SCENARIO,
        reportTargetDir = "build",
        testOutput = CourgetteTestOutput.CONSOLE,
        environmentInfo = "app=iOS test application; project_info=Courgette-JVM is awesome!",
        plugin = {CourgettePlugin.MOBILE_DEVICE_ALLOCATOR},
        mobileDeviceType = MobileDeviceType.SIMULATOR,
        mobileDevice = {
                "iPhone 15",
                "iPhone 15 Plus",
                "iPhone 15 Pro Max",
        },
        cucumberOptions = @CucumberOptions(
                features = "src/test/resources/features",
                glue = "steps",
                tags = "@ios",
                publish = true,
                plugin = {
                        "pretty",
                        "json:build/cucumber-report/cucumber.json",
                        "html:build/cucumber-report/cucumber.html"}
        ))
```

Features:
```
@ios
Feature: Test iOS application

  @Test1 @UploadPhotos
  Scenario: Verify application alert on device 1
    Given I launch the app
    When I show the alert
    Then I verify the alert shows this alert is so cool.
    And I accept the alert

  @Test2
  Scenario: Verify application alert on device 2
    Given I launch the app
    When I show the alert
    Then I verify the alert shows this alert is so cool.
    And I accept the alert

  @Test3
  Scenario: Verify application alert on device 3
    Given I launch the app
    When I show the alert
    Then I verify the alert shows this alert is so cool.
    And I accept the alert

  @Test4 @UploadPhotos
  Scenario: Verify application alert on device 4
    Given I launch the app
    When I show the alert
    Then I verify the alert shows this alert is so cool.
    And I accept the alert
```

**Comment by prashant-ramcharan** [comment]
I'm using the same project as shown in the videos above so cannot replicate your issue. 

Are you able to provide any example project for me to reproduce this?  I cannot reproduce this with the Courgette example project as shown in the videos.

> Case 1, when using either MobileDeviceType.SIMULATOR or MobileDeviceType.SIMULATOR_AND_REAL_DEVICE and running with 3 simulators or 3 simulators + 1 real device:

Could you please share your Courgette html report? This should show any exceptions thrown in the test.

**Comment by prashant-ramcharan** [comment]
@fatahillahardhi  In the mean time, I will release a new Courgette option to log the run details for each parallel run.

**Comment by prashant-ramcharan** [comment]
@fatahillahardhi I just released Courgette `6.12.0` with a new courgette option `generateCourgetteRunLog`

Could you try updating your runner and include the following option?

````java
generateCourgetteRunLog = true
````

At the end of the test run you should have a courgette run log saved to your report directory.

Please share the run log and Courgette html report if you still want me to look into this.

**Comment by fatahillahardhi** [comment]
hi @prashant-ramcharan , sorry for the wait, I can only get back into this project now.

I have reattempted with some cases with version `6.12.0`

**Case 1: Simulator & Real Devices using realMobileDeviceTag @UploadPhotos**
- 2 Simulators and 1 Real Device
- Initially, only the 2 simulators with tags not containing @UploadPhotos are executed in parallel.
- Once these 2 simulators are running, the real device does not react at all, more like waiting for the scenario without @UploadPhotos to finish.
- When the @UploadPhotos scenario finishes, with only scenarios containing @UploadPhotos remaining, the Real Device starts running, resulting in no parallel execution (with Real Devices). Is this expected?

**Case 2: Simulator & Real Devices not using realMobileDeviceTag @UploadPhotos**
- 2 Simulators and 1 Real Device
- Only simulators run in parallel; the real device does not run.
- Everything else remains as expected.

Here's the report: [Link](https://drive.google.com/drive/u/0/folders/19hj6jDSgQr-EKwfGMuRGg-GbrEdyepzU)
